File: project_euler/Problem077/Problem077.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem077() {
    target := 5000
    primes := make([]int, 0)
    primes = append(primes, 2)
    primes = append(primes, 3)
    result := 11
    for getWays(result, primes) <= target {
        result++
    }
    fmt.Println(result)
}

func getWays(target int, primes []int) int {
    index := primes[len(primes) - 1] + 2
    for index <= target {
        if isPrime(index) {
            primes = append(primes, index)
        }
        index += 2
    }

    ways := make([]int, target + 1)
    for i := 1; i <= target; i++ {
        ways[i] = 0
    }
    ways[0] = 1
    for _, prime := range primes {
        for i := prime; i <= target; i++ {
            ways[i] += ways[i - prime]
        }
    }
    return ways[target]
}

func isPrime(value int) bool {
    if 1 != value % 6 && 5 != value % 6 {
        return 2 == value || 3 == value
    }

    border := int(math.Sqrt(float64(value)))
    for i := 5; i <= border; i += 6 {
        if 0 == value % i || 0 == value % (i + 2) {
            return false
        }
    }
    return 1 != value
}

File: project_euler/Problem077/Problem077.py
import math
from typing import List

class Problem077:
    def solution(self) -> None:
        target, primes, result = 5000, [2, 3], 11
        while self.__get_ways(result, primes) <= target:
            result += 1
        print(result)

    def __get_ways(self, target: int, primes: List[int]) -> int:
        index = primes[-1] + 2
        while index <= target:
            if self.__is_prime(index):
                primes.append(index)
            index += 2

        ways = [0 for _ in range(target + 1)]
        ways[0] = 1
        for prime in primes:
            for i in range(prime, target + 1):
                ways[i] += ways[i - prime]
        return ways[target]

    def __is_prime(self, value: int) -> bool:
        if 1 != value % 6 and 5 != value % 6:
            return 2 == value or 3 == value

        border = int(math.sqrt(value))
        for i in range(5, border + 1, 6):
            if not value % i or not value % (i + 2):
                return False
        return 1 != value
